envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -0,0 +1,102 @@
+                        "It means that noone will be able to contact that endpoint.")
+                return@forEach
+            }
+
+            val policy = Policy.newBuilder()
+            val clientName = incomingEndpoint.clients.joinToString(",")
+
+            val pathPermission = Permission.newBuilder().setHeader(getPathMatcher(incomingEndpoint))
+            val methodPermissions = Permission.newBuilder()
+            val combinedPermissions = Permission.newBuilder()
+
+            val principals = incomingEndpoint.clients.map(this::mapClientToPrincipal)
+            policy.addAllPrincipals(principals)
+
+            if (incomingEndpoint.methods.isNotEmpty()) {
+                val methodPermissionSet = Permission.Set.newBuilder()
+                val methodPermissionsList = incomingEndpoint.methods.map(this::mapMethodToHeaderMatcher)
+                methodPermissionSet.addAllRules(methodPermissionsList)
+                methodPermissions.setOrRules(methodPermissionSet.build())
+                combinedPermissions.setAndRules(
+                        Permission.Set.newBuilder().addAllRules(listOf(
+                                pathPermission.build(),
+                                methodPermissions.build()
+                        ))
+                )
+            } else {
+                combinedPermissions.setAndRules(
+                        Permission.Set.newBuilder().addAllRules(listOf(pathPermission.build()))
+                )
+            }
+
+            policy.addPermissions(combinedPermissions)
+            clientToPolicy[clientName] = policy.build()
+        }
+
+        val rbac = RBAC.newBuilder()
+                .setAction(RBAC.Action.ALLOW)
+                .putAllPolicies(clientToPolicy)
+                .build()
+        return rbac
+    }
+
+    private fun mapMethodToHeaderMatcher(method: String): Permission {
+        val methodMatch = HeaderMatcher.newBuilder().setName(":method").setExactMatch(method).build()
+        return Permission.newBuilder().setHeader(methodMatch).build()
+    }
+
+    private fun mapClientToPrincipal(client: String): Principal {
+        val clientMatch = HeaderMatcher.newBuilder()
+                .setName(properties.clientIdentityHeader).setExactMatch(client).build()
+        return Principal.newBuilder().setHeader(clientMatch).build()
+    }
+
+    private fun getPathMatcher(incomingEndpoint: IncomingEndpoint): HeaderMatcher {
+        return when (incomingEndpoint.pathMatchingType) {
+            PathMatchingType.PATH -> HeaderMatcher.newBuilder()
+                    .setName(":path").setExactMatch(incomingEndpoint.path).build()
+            PathMatchingType.PATH_PREFIX -> HeaderMatcher.newBuilder()
+                    .setName(":path").setPrefixMatch(incomingEndpoint.path).build()
+        }
+    }
+
+    fun createHttpFilter(group: Group): HttpFilter? {
+        return if (properties.enabled) {
```

Comment:
How about setting shadowRules instead of null filter when RBAC is disabled?
```kotlin
    fun createHttpFilter(group: Group): HttpFilter? {
        val rules = getRules(group.serviceName, group.proxySettings.incoming)
        val rbacFilter = if (properties.enabled) {
            RBACFilter.newBuilder().setRules(rules).build()
        } else {
            RBACFilter.newBuilder().setShadowRules(rules).build()
        }
        return HttpFilter.newBuilder().setName("envoy.filters.http.rbac").setTypedConfig(Any.pack(rbacFilter)).build()
    }
```

Reply:
Hmm... good idea but maybe it should be a separate setting like: enabled, shadow, disabled. WDYT @MarcinFalkowski ?